src/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/HttpHandlerDiagnosticListener.cs

```diff
@@ -0,0 +1,370 @@
+                        {
+                            newArrayList.Add(entry);
+                        }
+                    }
+                    s_writeListField.SetValue(value, newArrayList);
+                }
+
+                return base.Add(value);
+            }
+        }
+
+        /// <summary>
+        /// Helper class used for Connection.m_WriteList. The goal here is to
+        /// intercept all new HttpWebRequest objects being added to Connection.m_WriteList
+        /// and notify the listener about the HttpWebRequest that's about to send a request.
+        /// It also intercepts all HttpWebRequest objects that are about to get removed from
+        /// Connection.m_WriteList as they have completed the request.
+        /// </summary>
+        private class HttpWebRequestArrayList : ArrayList
+        {
+            public override int Add(object value)
+            {
+                HttpWebRequest request = value as HttpWebRequest;
+                if (request != null)
+                {
+                    s_instance.RaiseRequestEvent(request);
+                }
+
+                return base.Add(value);
+            }
+
+            public override void RemoveAt(int index)
+            {
+                HttpWebRequest request = base[index] as HttpWebRequest;
+                if (request != null)
+                {
+                    HttpWebResponse response = s_httpResponseAccessor(request);
+                    if (response != null)
+                    {
+                        s_instance.RaiseResponseEvent(request, response);
+                    }
+                }
+
+                base.RemoveAt(index);
+            }
+        }
+
+#endregion
+
+#region private methods
+
+        /// <summary>
+        /// Private constructor. Make sure every caller has to create this object via the Initialize
+        /// method.
+        /// </summary>
+        private HttpHandlerDiagnosticListener() : base(DiagnosticListenerName)
+        {
+        }
+
+        private void RaiseRequestEvent(HttpWebRequest request)
+        {
+            Guid loggingRequestId = Guid.Empty;
+
+            // If System.Net.Http.Request is on, raise the event
+            if (this.IsEnabled(RequestWriteName))
+            {
+                long timestamp = Stopwatch.GetTimestamp();
+                loggingRequestId = Guid.NewGuid();
+                this.Write(RequestWriteName,
+                    new
+                    {
+                        Request = request,
+                        //LoggingRequestId = loggingRequestId,
+                        Timestamp = timestamp
+                    }
+                );
+            }
+        }
+
+        private void RaiseResponseEvent(HttpWebRequest request, HttpWebResponse response)
+        {
+            Guid loggingRequestId = Guid.Empty;
+
+            // TODO: Figure what's the right way to detect error conditions here.
+            if (response.StatusCode >= (HttpStatusCode)400)
```

Comment:
I don't think Exception event should be sent on >=400 status code.
Consumer can always use response object and status code.